include/ccf/indexing/strategies/seqnos_by_key_bucketed.h

```diff
@@ -0,0 +1,373 @@
+      current_results;
+
+    // Maintain an LRU of old, requested buckets, which are asynchronously
+    // fetched from disk
+    using BucketKey = std::pair<kv::untyped::SerialisedEntry, Range>;
+    // First element is a handle while result is being fetched. Second is parsed
+    // result, after fetch completes, at which point the first is set to nullptr
+    using BucketValue = std::pair<FetchResultPtr, SeqNoCollection>;
+    LRU<BucketKey, BucketValue> old_results;
+
+    AbstractLFSAccess& lfs_access;
+
+    LFSContents serialise(SeqNoCollection&& seqnos)
+    {
+      LFSContents blob;
+
+      {
+        // Write number of seqnos
+        const auto orig_size = blob.size();
+        blob.resize(orig_size + sizeof(seqnos.size()));
+        auto data = blob.data() + orig_size;
+        auto size = blob.size() - orig_size;
+        serialized::write(data, size, seqnos.size());
+      }
+
+      // Write each seqno
+      for (const auto& seqno : seqnos)
+      {
+        const auto orig_size = blob.size();
+        blob.resize(orig_size + sizeof(seqno));
+        auto data = blob.data() + orig_size;
+        auto size = blob.size() - orig_size;
+        serialized::write(data, size, seqno);
+      }
+
+      return blob;
+    }
+
+    SeqNoCollection deserialise(const LFSContents& raw, bool& corrupt)
+    {
+      corrupt = false;
+      try
+      {
+        auto data = raw.data();
+        auto size = raw.size();
+
+        // Read number of seqnos
+        const auto seqno_count = serialized::read<size_t>(data, size);
+        SeqNoCollection seqnos;
+
+        for (auto j = 0u; j < seqno_count; ++j)
+        {
+          seqnos.insert(serialized::read<ccf::SeqNo>(data, size));
+        }
+
+        if (size != 0)
+        {
+          LOG_TRACE_FMT("{} bytes remaining after deserialisation", size);
+          corrupt = true;
+          return {};
+        }
+
+        return seqnos;
+      }
+      // Catch errors thrown by serialized::read
+      catch (const std::logic_error& e)
+      {
+        corrupt = true;
+        return {};
+      }
+    }
+
+    LFSKey get_blob_name(const BucketKey& bk)
+    {
+      const auto hex_key = ds::to_hex(bk.first.begin(), bk.first.end());
+      const auto& range = bk.second;
+      return fmt::format(
+        "{}: {} -> {} for {}", get_name(), range.first, range.second, hex_key);
+    }
+
+    void store_to_disk(
+      const kv::untyped::SerialisedEntry& k,
+      const Range& range,
+      SeqNoCollection&& seqnos)
+    {
+      const BucketKey bucket_key{k, range};
+      const auto blob_key = get_blob_name(bucket_key);
+      lfs_access.store(blob_key, serialise(std::move(seqnos)));
+    }
+
+    Range get_range_for(ccf::SeqNo seqno) const
+    {
+      const auto begin = (seqno / seqnos_per_bucket) * seqnos_per_bucket;
+      const auto end = begin + seqnos_per_bucket;
+      return {begin, end};
+    }
+
+    void visit_entry(
```

Comment:
This is the write-side of the new `Strategy`. Each time the `Indexer` gives it an entry (a write to the target table), we:
- Find the bucket to store this result in. This is one-bucket-per-key, and a caller-configured range of seqnos in each bucket.
- If we have a current bucket for this key, but from an old range, then dump that bucket to disk
  - That involves find a unique name for the bucket, serialising the contents, and handing it off to the LFS system for encryption.
  - Then building a new empty bucket for the correct range.
- Record the given seqno in this bucket.